pkg/comp-functions/functions/vshnkeycloak/deploy.go

```diff
@@ -0,0 +1,373 @@
+	}
+
+	if comp.Spec.Parameters.Service.PostgreSQLParameters != nil {
+		err := mergo.Merge(params, comp.Spec.Parameters.Service.PostgreSQLParameters, mergo.WithOverride)
+		if err != nil {
+			return err
+		}
+
+		// Mergo currently has a bug with merging bools: https://github.com/darccio/mergo/issues/249
+		// It's not possible to override true with false, so it won't merge this if the users disables it.
+		if !comp.Spec.Parameters.Service.PostgreSQLParameters.Backup.DeletionProtection {
+			params.Backup.DeletionProtection = false
+		}
+	}
+
+	pg := &vshnv1.XVSHNPostgreSQL{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: comp.GetName() + pgInstanceNameSuffix,
+		},
+		Spec: vshnv1.XVSHNPostgreSQLSpec{
+			Parameters: *params,
+			ResourceSpec: xpv1.ResourceSpec{
+				WriteConnectionSecretToReference: &xpv1.SecretReference{
+					Name:      pgSecretName,
+					Namespace: comp.GetInstanceNamespace(),
+				},
+			},
+		},
+	}
+
+	err := common.CreateNetworkPolicy([]string{comp.GetInstanceNamespace()}, pg.GetInstanceNamespace(), pg.GetName()+"-keycloak", svc)
+	if err != nil {
+		return err
+	}
+
+	return svc.SetDesiredComposedResource(pg)
+}
+
+func addRelease(ctx context.Context, svc *runtime.ServiceRuntime, comp *vshnv1.VSHNKeycloak, adminSecret string) error {
+	release, err := newRelease(ctx, svc, comp, adminSecret)
+	if err != nil {
+		return err
+	}
+
+	return svc.SetDesiredComposedResource(release)
+}
+
+func getResources(ctx context.Context, svc *runtime.ServiceRuntime, comp *vshnv1.VSHNKeycloak) (common.Resources, error) {
+	plan := comp.Spec.Parameters.Size.GetPlan(svc.Config.Data["defaultPlan"])
+
+	resources, err := utils.FetchPlansFromConfig(ctx, svc, plan)
+	if err != nil {
+		err = fmt.Errorf("cannot fetch plans from the composition config, maybe they are not set: %w", err)
+		return common.Resources{}, err
+	}
+
+	res := common.GetResources(&comp.Spec.Parameters.Size, resources)
+
+	return res, nil
+}
+
+func newValues(ctx context.Context, svc *runtime.ServiceRuntime, comp *vshnv1.VSHNKeycloak, adminSecret string) (map[string]any, error) {
+	values := map[string]any{}
+
+	cd, err := svc.GetObservedComposedResourceConnectionDetails(comp.GetName() + pgInstanceNameSuffix)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := getResources(ctx, svc, comp)
+	if err != nil {
+		return nil, err
+	}
+
+	plan := comp.Spec.Parameters.Size.GetPlan(svc.Config.Data["defaultPlan"])
+	nodeSelector, err := utils.FetchNodeSelectorFromConfig(ctx, svc, plan, comp.Spec.Parameters.Scheduling.NodeSelector)
+	if err != nil {
+		return values, fmt.Errorf("cannot fetch nodeSelector from the composition config: %w", err)
+	}
+
+	fqdn := ""
+	strictHostname := false
+	if comp.Spec.Parameters.Service.FQDN != "" {
+		fqdn = comp.Spec.Parameters.Service.FQDN
+		strictHostname = true
+	}
+
+	extraEnv, err := toYaml([]map[string]any{
+		{
+			"name": "JAVA_OPTS",
+			"value": `-XX:+UseContainerSupport
+-XX:MaxRAMPercentage=50.0
+-Djava.net.preferIPv4Stack=true
```

Comment:
Most probably not required anymore >= v22. https://github.com/projectsyn/component-keycloak/issues/231